packages/core/src/state-management/state-manager.ts

```diff
@@ -324,17 +325,31 @@ export class StateManager {
   /**
    * Request anchor for the latest stream state
    */
-  anchor(state$: RunningState): Subscription {
+  async anchor(state$: RunningState): Promise<void> {
     if (!this.anchorService) {
       throw new Error(
         `Anchor requested for stream ${state$.id.toString()} but anchoring is disabled`
       )
     }
     if (state$.value.anchorStatus == AnchorStatus.ANCHORED) {
-      return Subscription.EMPTY
+      return
     }
-    const anchorStatus$ = this.anchorService.requestAnchor(state$.id, state$.tip)
-    return this._processAnchorResponse(state$, anchorStatus$)
+
+    return new Promise((resolve) => {
+      this._saveAnchorRequestForState(state$).then(() => {
+        const anchorStatus$ = this.anchorService.requestAnchor(state$.id, state$.tip)
+        let resolved = false
+        const observable$ = anchorStatus$.pipe(
+          tap((value) => {
+            if (!resolved && value.status === AnchorStatus.PENDING) {
```

Comment:
what is this doing?  Why do we need to wait for PENDING here?

Reply:
Can this whole block just be
```
await this._saveAnchorRequestForState(state$)
const anchorStatus$ = this.anchorService.requestAnchor(state$.id, state$.tip)
return this._processAnchorResponse(state$, anchorStatus$)
```
?

Reply:
> Can this whole block just be
```
await this._saveAnchorRequestForState(state$)
const anchorStatus$ = this.anchorService.requestAnchor(state$.id, state$.tip)
return this._processAnchorResponse(state$, anchorStatus$)
```

We don't want to do this like that, because then the return type of this method would be `Promise<Subscription>` and we don't want to mix promises with `rxjs` types.

If we don't want to wait for PENDING HERE, I think we can just change this code to:
```
await this._saveAnchorRequestForState(state$)
const anchorStatus$ = this.anchorService.requestAnchor(state$.id, state$.tip)
this._processAnchorResponse(state$, anchorStatus$)
```

, but I'm not sure, if it doesn't break some assumptions in tests or in other logic elsewhere in `js-ceramic`, tbh, I'd have to check.

Reply:
Having said all of the above, here are the available anchor states right now:
```
export enum AnchorStatus {
  NOT_REQUESTED,
  PENDING,
  PROCESSING,
  ANCHORED,
  FAILED,
}
```

What anchor state should the stream have, in your opinion, @stbrody , after `anchor(...)` is called, if it doesn't wait for `PENDING`?

Reply:
the current function gives no guarantee as to the anchor status after `anchor()` returns, it only guarantees that the anchor process has been kicked off.  

I guess the promise-y way to do what the current code does with a Subscription would be to return a Promise that resolves when the anchor request either succeeds or definitively fails.

Did you give up on trying to make this work in a Subscription-friendly way?  I'd be curious for thoughts from @ukstv on what the right behavior here should be.

Reply:
Can you explain what you mean by "Subscription-friendly way", @stbrody ?

The goal of this PR is to add a single async step, before an anchor is requested. Currently in `develop` we use a sync `anchor(...)` method which returns an instance of `Subscription`

So, the way I see it, we have two options here:
1. Keep the `anchor(...)` method as is, but add the new async call in before every call to `anchor(...`)
2. Make `anchor(...)` async to do the additional async work inside it, so that we make sure it's always done when `anchor(...)` is called

If we go with option 2 (which is what I decided to do in this PR), we have yet two more options:
a. Maintain the current contract of `anchor(...)` and continue returning an instance of `Subscription`, just make it async. This would lead to a return type of `Promise<Subscription>`, though, which we pointed out we didn't want in this [comment](https://github.com/ceramicnetwork/js-ceramic/pull/2566#discussion_r1029340359)
b. Slightly change the behaviour of the system by not returning an instance of `Subscription` anymore. Instead we change the return type of `anchor(...)` to `Promise<void>`, so it's not possible to add handlers to this subscription anymore (something we seem only to be using in tests right now, e.g. [here](https://github.com/ceramicnetwork/js-ceramic/blob/develop/packages/core/src/__tests__/state-manager.test.ts#L92) )

Do these options make sense to you, @stbrody ? Do you see any other options that we missed? What would you propose?

Reply:
@ukstv , what do you think about it :) ?

Reply:
@stbrody I tend to align with @pawartur here, and I agree it is a bit tricky.

Original `anchor` was really about kicking off the process. Now we have some part (=store the anchor to the store) that must succeed before we go any further. Thus, here we ensure that we store the anchor, and then kick off the polling process.

`anchor` method is called from two places: `Ceramic#requestAnchor` and `StateManager#applyWriteOps`. Both run as part of incoming HTTP request processing. Storing an anchor resembles pinning and storing the state, so here we continue a tradition of making storage awaitable.

As for this: "... to return a Promise that resolves when the anchor request either succeeds or definitively fails", the code here goes exactly as you wrote. If we have stored the anchor request in the store, we can be sure of its future fate.